specs/vep-2448-vdk-run-logs-simplified-and-readable/README.md

```diff
@@ -366,10 +366,290 @@ or slowed down due to progress bar updates. This is really relevant for
 ingestion.
 
 Also as we need progress indicators in multiple places likely we need a common
-encapsulation for that - in a form of a python module or plugin. The progress
+encapsulation (or abstraction) for that - in a form of a python module or plugin. The progress
 indicator should provide a notification callback mechanism or similar so it's
 integratable with Notebook
 
+For more detailed understanding of the user workflows [this research documented is recommended](https://github.com/vmware/versatile-data-kit/wiki/Research:-Progress-Indicators)
+
+#### Technical Requirements
+
+* The progress abstraction should support multi step operations. Different steps are started and ended multiple times one after another in a job.
+* The abstraction should support hierarchical operations. A Step can start other long running operations (e.g a DAG sub-jobs)
+* The abstraction should support operations running concurrently (that is multiple tasks are started and keep progressing concurrently - in the same thread).
+  For example, this requirement comes primarily because of DAGs. Multiple sub-jobs started in a DAG can run concurrently (no multi-threading needed here though)
+* The abstraction should support operations running in background threads. When you send data for ingestion, that data is processed in background while the job continue with other operations. The progress track can be updated from multiple threads as they ingest data
+* The abstraction should allow for the total to change. E.g when we are ingesting data the total number of objects being ingested is changed as user adds new and new ones.
+* Performance overhead should be minimal as this may be called millions of times (e.g for each ingested row in a payload)
+* (Optional) The abstraction can be extended to collect metrics  about the state of operations running by the framework (bandwidth, duration, failed requests, etc).
+  This is currently used by vdk-core ingestion logic.
+
+
+##### Core Components
+
+###### ProgressTracker
+The core component that handles tracking of the progress for operations.
+This component will be part of vdk-core and will offer foundational tracking functionalities based on the core requirements.
+
+###### API Methods
+
+- create_sub_tracker(): To create a new sub-tracker for nested tasks.
+- add(n): To add n number of new iterations (items) for tracking.
+- update(n): To update the progress by n iterations.
+
+###### Workflow
+- On CLI execution start we will create progress tracker and store it in [CoreContext](https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-core/src/vdk/internal/core/context.py).
```

Comment:
This sounds like an observer to me again. The only difference I see down the line is we want the observers to also form a tree, because updating a child observer might also cause the parent to update.